server/mediaserver.go

```diff
@@ -395,9 +413,27 @@ func gotRTMPStreamHandler(s *LivepeerServer) func(url *url.URL, rtmpStrm stream.
 							}
 							newSeg := &stream.HLSSegment{SeqNo: seg.SeqNo, Name: parseSegName(url), Data: data, Duration: seg.Duration}
 							s.LivepeerNode.VideoSource.InsertHLSSegment(sid, newSeg)
+
+							segHashes[i] = crypto.Keccak256(data)
```

Comment:
This assignment still needs to be protected by SegHashLock, since it could be modified concurrently by any of the goroutines running `dlFunc`

Also, without actually benchmarking `crypto.Keccak256` , my guess is that it might take a little while (eg, more than a few ms) for a large segment. Maybe it'd be good to do this outside the lock, eg
```go
hash := crypto.Keccak256(data)
segHashLock.Lock()
segHashes[i] = hash
segHashLock.Unlock()
```

That we're actually sharing the same lock as the condvar is OK in this case; you'd just be acquiring and releasing the lock twice (once on assignment, once when the function exits to decrement the counter). There are ways avoid doing that, but we don't need to optimize that much here.